metricbeat/module/beat/beat.go

```diff
@@ -79,35 +79,27 @@ type State struct {
 }
 
 // GetInfo returns the data for the Beat's / endpoint.
-func GetInfo(m *MetricSet) (*Info, error) {
+func GetInfo(m *MetricSet) (Info, error) {
+	info := Info{}
 	content, err := fetchPath(m.HTTP, "/")
 	if err != nil {
-		return nil, err
+		return info, err
 	}
 
-	info := &Info{}
 	err = json.Unmarshal(content, &info)
-	if err != nil {
-		return nil, err
-	}
-
-	return info, nil
+	return info, err
 }
```

Comment:
A couple of small suggestions/improvements:
1. It's better to explicitly return the zero-value struct (`Info{}`) on errors
2. You can instantiate `info` right before it's used and also return a zero-value if unmarshal fails, so there is no chance of the returned value containing partial information. There are some edge cases where an error is returned and partial information is unmarshalled into the struct.

```suggestion
func GetInfo(m *MetricSet) (Info, error) {
	content, err := fetchPath(m.HTTP, "/")
	if err != nil {
		return Info{}, err
	}

	info := Info{}
	if err:= json.Unmarshal(content, &info); err != nil {
		return Info{}, err
	}

	return info, err
}
```

Reply:
Curious if there are benefits other than consistent behavior ? Since the caller is expected to check for errors before proceeding is there any harm in returning partial data

Reply:
Consistent behaviour is definitely a key point here, on errors is common to return a zero-value, making it explicit also helps readability.

Even if the caller does not properly handle the error, at least this method is not returning a partial value.